include/tvm/tir/expr.h

```diff
@@ -610,6 +610,14 @@ class BufferLoadNode : public PrimExprNode {
   /*! \brief The indices location to be loaded. */
   Array<PrimExpr> indices;
 
+  /*! \brief Set the dtype based on the buffer/indices
+   *
+   * Usually, this will be the same dtype as the buffer.  This may
+   * have a different number of lanes than the buffer's dtype if index
+   * values have more than 1 lane.
+   */
+  void LegalizeDtype();
```

Comment:
This is primarily to allow for shared implementations between the `BufferStore` and `BufferLoad` visitors, without needing a dedicated node to represent the buffer access.  See [this comment](https://github.com/apache/tvm-rfcs/pull/42#issuecomment-963314748) for the discussion and an example.

In this case, the `LegalizeDtype` function is called from the constructor, but is also callable after `CopyOnWrite`+mutation.  (e.g. [link](https://github.com/Lunderberg/tvm/blob/1dd14389a/src/tir/transforms/storage_rewrite.cc#L1391))

Reply:
DType(to be consistent with other places, as DType is short for DataType). please add comment that this is only intended to be called from ctor and after COW. Consider add an underscore after function to indicate the intended rare behavior. Even better, we might want to only make it private and expose to limited locations via friend, but this can be left as a TODO